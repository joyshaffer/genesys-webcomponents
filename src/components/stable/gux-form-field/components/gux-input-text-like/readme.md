# gux-input-text-like

<!-- Auto Generated Below -->


## Properties

| Property    | Attribute   | Description | Type      | Default     |
| ----------- | ----------- | ----------- | --------- | ----------- |
| `clearable` | `clearable` |             | `boolean` | `undefined` |


## Slots

| Slot      | Description                                                             |
| --------- | ----------------------------------------------------------------------- |
| `"input"` | Required slot for input[type="text" \| type="email" \| type="password"] |


## Dependencies

### Used by

 - [gux-dropdown](../../../gux-dropdown)
 - [gux-form-field](../..)

### Depends on

- [gux-icon](../../../gux-icon)

### Graph
```mermaid
graph TD;
  gux-input-text-like --> gux-icon
  gux-dropdown --> gux-input-text-like
  gux-form-field --> gux-input-text-like
  style gux-input-text-like fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
