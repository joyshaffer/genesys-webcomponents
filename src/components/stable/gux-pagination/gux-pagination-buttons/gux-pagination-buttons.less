@import (reference) '../../../../style/color-palette.less';

gux-pagination-buttons {
  color: @gux-black-50;

  .gux-pagination-buttons-container {
    display: flex;
    flex-direction: row;
    flex-wrap: nowrap;
    align-content: stretch;
    align-items: center;
    justify-content: flex-end;

    & > div {
      flex: 0 1 auto;
      align-self: auto;
      order: 0;
    }

    .gux-pagination-buttons-group {
      white-space: nowrap;

      gux-button-slot-beta:not(:first-of-type) {
        margin-left: 8px;
      }
    }

    .gux-pagination-buttons-input-container {
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      align-content: stretch;
      align-items: center;
      justify-content: center;
      margin: 0 16px;

      & > div {
        flex: 0 1 auto;
        align-self: auto;
        order: 0;
        white-space: nowrap;
      }

      .gux-pagination-buttons-input {
        width: 60px;
        margin: 0 8px;

        input {
          width: inherit;
          text-align: center;
        }
      }
    }

    .gux-pagination-buttons-list-container {
      display: flex;
      flex-direction: row;
      flex-wrap: nowrap;
      align-content: stretch;
      align-items: center;
      justify-content: center;

      .gux-pagination-buttons-list-current,
      .gux-pagination-buttons-list-target {
        height: 32px;
        padding: 0 8px;
        margin: 0;
        background: inherit;
        border: none;

        &:focus-visible {
          border: 3px solid @gux-blue-90;
          border-radius: 5px;
          outline: none;
        }
      }

      .gux-pagination-buttons-list-current {
        font-weight: 600;
      }

      .gux-pagination-buttons-list-target {
        color: @gux-blue-60;
        cursor: pointer;
      }
    }

    .gux-pagination-buttons-spacer {
      width: 8px;
    }
  }
}
