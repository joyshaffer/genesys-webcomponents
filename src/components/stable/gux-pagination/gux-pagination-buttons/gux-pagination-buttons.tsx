import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop
} from '@stencil/core';

import { buildI18nForComponent, GetI18nValue } from '../../../../i18n';
import { GuxPaginationLayout } from '../gux-pagination.types';

import paginationResources from './i18n/en.json';
import { GuxPaginationButtonsService } from './gux-pagination-button.service';

@Component({
  styleUrl: 'gux-pagination-buttons.less',
  tag: 'gux-pagination-buttons'
})
export class GuxPaginationButtons {
  @Element()
  private root: HTMLElement;

  private i18n: GetI18nValue;

  @Prop()
  currentPage: number;

  @Prop()
  totalPages: number;

  @Prop()
  layout: GuxPaginationLayout = 'full';

  @Event({ bubbles: false })
  private internalcurrentpagechange: EventEmitter<number>;

  private get onFirstPage(): boolean {
    return this.currentPage <= 1;
  }

  private get onLastPage(): boolean {
    return this.currentPage >= this.totalPages;
  }

  private handleClickFirst(): void {
    this.internalcurrentpagechange.emit(1);
  }

  private handleClickPrevious(): void {
    this.internalcurrentpagechange.emit(this.currentPage - 1);
  }

  private handleClickNext(): void {
    this.internalcurrentpagechange.emit(this.currentPage + 1);
  }

  private handleClickLast(): void {
    this.internalcurrentpagechange.emit(this.totalPages);
  }

  private handleClickPage(pageNumber: number): void {
    this.internalcurrentpagechange.emit(pageNumber);
  }

  private getPageListEnteries(
    currentPage: number,
    totalPages: number
  ): JSX.Element[] {
    return GuxPaginationButtonsService.getPageList(
      currentPage,
      totalPages
    ).reduce((acc, cv) => {
      if (cv.current) {
        return acc.concat(
          (
            <button class="gux-pagination-buttons-list-current">
              {cv.display}
            </button>
          ) as JSX.Element
        );
      }

      return acc.concat(
        (
          <button
            class="gux-pagination-buttons-list-target"
            onClick={() => this.handleClickPage(cv.pageNumber)}
          >
            {cv.display}
          </button>
        ) as JSX.Element
      );
    }, [] as JSX.Element[]);
  }

  private getSmallPagePicker(): JSX.Element {
    return (<div class={'gux-pagination-buttons-spacer'}></div>) as JSX.Element;
  }

  private getFullPagePicker(): JSX.Element {
    return (
      <div class="gux-pagination-buttons-list-container">
        {this.getPageListEnteries(this.currentPage, this.totalPages)}
      </div>
    ) as JSX.Element;
  }

  private getPagePicker(layout: GuxPaginationLayout): JSX.Element {
    if (layout === 'simple') {
      return this.getSmallPagePicker();
    }

    return this.getFullPagePicker();
  }

  async componentWillLoad(): Promise<void> {
    this.i18n = await buildI18nForComponent(this.root, paginationResources);
  }

  render(): JSX.Element {
    return (
      <div class={`gux-pagination-buttons-container gux-${this.layout}`}>
        <div class="gux-pagination-buttons-group">
          <gux-button-slot-beta accent="secondary">
            <button
              title={this.i18n('first')}
              disabled={this.onFirstPage}
              onClick={this.handleClickFirst.bind(this)}
            >
              <gux-icon decorative icon-name="chevron-double-left"></gux-icon>
            </button>
          </gux-button-slot-beta>
          <gux-button-slot-beta accent="secondary">
            <button
              title={this.i18n('previous')}
              disabled={this.onFirstPage}
              onClick={this.handleClickPrevious.bind(this)}
            >
              <gux-icon decorative icon-name="chevron-small-left"></gux-icon>
            </button>
          </gux-button-slot-beta>
        </div>

        {this.getPagePicker(this.layout)}

        <div class="gux-pagination-buttons-group">
          <gux-button-slot-beta accent="secondary">
            <button
              title={this.i18n('next')}
              disabled={this.onLastPage}
              onClick={this.handleClickNext.bind(this)}
            >
              <gux-icon decorative icon-name="chevron-small-right"></gux-icon>
            </button>
          </gux-button-slot-beta>
          <gux-button-slot-beta accent="secondary">
            <button
              title={this.i18n('last')}
              disabled={this.onLastPage}
              onClick={this.handleClickLast.bind(this)}
            >
              <gux-icon decorative icon-name="chevron-double-right"></gux-icon>
            </button>
          </gux-button-slot-beta>
        </div>
      </div>
    ) as JSX.Element;
  }
}
