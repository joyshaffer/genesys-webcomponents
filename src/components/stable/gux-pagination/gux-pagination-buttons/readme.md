# gux-pagination-buttons

An internal component used by the gux-pagination component.


<!-- Auto Generated Below -->


## Properties

| Property      | Attribute      | Description | Type                | Default     |
| ------------- | -------------- | ----------- | ------------------- | ----------- |
| `currentPage` | `current-page` |             | `number`            | `undefined` |
| `layout`      | `layout`       |             | `"full" \| "small"` | `'full'`    |
| `totalPages`  | `total-pages`  |             | `number`            | `undefined` |


## Events

| Event                       | Description | Type                  |
| --------------------------- | ----------- | --------------------- |
| `internalcurrentpagechange` |             | `CustomEvent<number>` |


## Dependencies

### Used by

 - [gux-pagination](..)

### Depends on

- [gux-button-slot-beta](../../../beta/gux-button-slot)
- [gux-icon](../../gux-icon)

### Graph
```mermaid
graph TD;
  gux-pagination-buttons --> gux-button-slot-beta
  gux-pagination-buttons --> gux-icon
  gux-pagination --> gux-pagination-buttons
  style gux-pagination-buttons fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
