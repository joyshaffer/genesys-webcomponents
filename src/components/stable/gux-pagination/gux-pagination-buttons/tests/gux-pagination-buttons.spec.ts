import { newSpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxPaginationButtons } from '../gux-pagination-buttons';
import { GuxButton } from '../../../gux-button/gux-button';
import { GuxInputTextLike } from '../../../gux-form-field/components/gux-input-text-like/gux-input-text-like';

const components = [GuxPaginationButtons, GuxButton, GuxInputTextLike];
const language = 'en';

describe('gux-pagination-item-counts', () => {
  beforeEach(() => {
    (
      global as NodeJS.Global & {
        MutationObserver: any;
      }
    ).MutationObserver = MutationObserver;
  });

  describe('#render', () => {
    [
      { totalPages: 0, currentPage: 0, layout: 'full' },
      { totalPages: 10, currentPage: 1, layout: 'full' },
      { totalPages: 10, currentPage: 5, layout: 'full' },
      { totalPages: 0, currentPage: 0, layout: 'small' },
      { totalPages: 10, currentPage: 1, layout: 'small' },
      { totalPages: 10, currentPage: 5, layout: 'small' }
    ].forEach(({ totalPages, currentPage, layout }, index) => {
      it(`should render as expected (${index + 1})`, async () => {
        const html = `
          <gux-pagination-buttons
            total-pages="${totalPages}"
            current-page="${currentPage}"
            layout="${layout}"
          >
          </gux-pagination-buttons>
        `;
        const page = await newSpecPage({ components, html, language });

        expect(page.rootInstance).toBeInstanceOf(GuxPaginationButtons);

        expect(page.root).toMatchSnapshot();
      });
    });
  });
});
