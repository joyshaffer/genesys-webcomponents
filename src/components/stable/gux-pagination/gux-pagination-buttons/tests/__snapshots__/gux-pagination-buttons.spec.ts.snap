// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-pagination-item-counts #render should render as expected (1) 1`] = `
<gux-pagination-buttons current-page="0" layout="full" total-pages="0">
  <div class="gux-full gux-pagination-buttons-container">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-list-container"></div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;

exports[`gux-pagination-item-counts #render should render as expected (2) 1`] = `
<gux-pagination-buttons current-page="1" layout="full" total-pages="10">
  <div class="gux-full gux-pagination-buttons-container">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-list-container">
      <button class="gux-pagination-buttons-list-current">
        1
      </button>
      <button class="gux-pagination-buttons-list-target">
        2
      </button>
      <button class="gux-pagination-buttons-list-target">
        3
      </button>
      <button class="gux-pagination-buttons-list-target">
        4
      </button>
      <button class="gux-pagination-buttons-list-target">
        5
      </button>
      <button class="gux-pagination-buttons-list-target">
        6
      </button>
      <button class="gux-pagination-buttons-list-target">
        7
      </button>
      <button class="gux-pagination-buttons-list-target">
        8
      </button>
      <button class="gux-pagination-buttons-list-target">
        9
      </button>
      <button class="gux-pagination-buttons-list-target">
        10
      </button>
    </div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;

exports[`gux-pagination-item-counts #render should render as expected (3) 1`] = `
<gux-pagination-buttons current-page="5" layout="full" total-pages="10">
  <div class="gux-full gux-pagination-buttons-container">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-list-container">
      <button class="gux-pagination-buttons-list-target">
        1
      </button>
      <button class="gux-pagination-buttons-list-target">
        2
      </button>
      <button class="gux-pagination-buttons-list-target">
        3
      </button>
      <button class="gux-pagination-buttons-list-target">
        4
      </button>
      <button class="gux-pagination-buttons-list-current">
        5
      </button>
      <button class="gux-pagination-buttons-list-target">
        6
      </button>
      <button class="gux-pagination-buttons-list-target">
        7
      </button>
      <button class="gux-pagination-buttons-list-target">
        8
      </button>
      <button class="gux-pagination-buttons-list-target">
        9
      </button>
      <button class="gux-pagination-buttons-list-target">
        10
      </button>
    </div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;

exports[`gux-pagination-item-counts #render should render as expected (4) 1`] = `
<gux-pagination-buttons current-page="0" layout="small" total-pages="0">
  <div class="gux-pagination-buttons-container gux-small">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-spacer"></div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;

exports[`gux-pagination-item-counts #render should render as expected (5) 1`] = `
<gux-pagination-buttons current-page="1" layout="small" total-pages="10">
  <div class="gux-pagination-buttons-container gux-small">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button disabled="" title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-spacer"></div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;

exports[`gux-pagination-item-counts #render should render as expected (6) 1`] = `
<gux-pagination-buttons current-page="5" layout="small" total-pages="10">
  <div class="gux-pagination-buttons-container gux-small">
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="First">
          <gux-icon decorative="" icon-name="chevron-double-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Previous">
          <gux-icon decorative="" icon-name="chevron-small-left"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
    <div class="gux-pagination-buttons-spacer"></div>
    <div class="gux-pagination-buttons-group">
      <gux-button-slot-beta accent="secondary">
        <button title="Next">
          <gux-icon decorative="" icon-name="chevron-small-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
      <gux-button-slot-beta accent="secondary">
        <button title="Last">
          <gux-icon decorative="" icon-name="chevron-double-right"></gux-icon>
        </button>
      </gux-button-slot-beta>
    </div>
  </div>
</gux-pagination-buttons>
`;
